pydocstringformatter/formatting.py

```diff
@@ -1,39 +1,74 @@
+import abc
 import re
 import tokenize
 
+# pylint: disable=too-few-public-methods
+from typing import List
 
-def _format_beginning_quotes(tokeninfo: tokenize.TokenInfo) -> tokenize.TokenInfo:
+
+class Formatter:
+    """Base class for docstring formatter"""
+
+    @abc.abstractmethod
+    def treat(self, tokeninfo: tokenize.TokenInfo) -> tokenize.TokenInfo:
+        """Return a modified token"""
+
+
+class StringFormatter(Formatter):
+    """Base class for formatter that only modify the string content"""
+
+    @abc.abstractmethod
+    def _treat_string(self, new_string: str) -> str:
+        """Will modify the string"""
+
+    def treat(self, tokeninfo: tokenize.TokenInfo) -> tokenize.TokenInfo:
+        return tokenize.TokenInfo(
+            tokeninfo.type,
+            self._treat_string(tokeninfo.string),
+            tokeninfo.start,
+            tokeninfo.end,
+            tokeninfo.line,
+        )
+
+
+class BeginningQuotesFormatter(StringFormatter):
     """Fix the position of the opening quotes"""
-    new_string = tokeninfo.string
-    if new_string[3] == "\n":
-        new_string = re.sub(r"\n *", "", new_string, 1)
-    return tokenize.TokenInfo(
-        tokeninfo.type, new_string, tokeninfo.start, tokeninfo.end, tokeninfo.line
-    )
-
-
-def _format_multiline_ending_quotes(
-    tokeninfo: tokenize.TokenInfo,
-) -> tokenize.TokenInfo:
-    """Fix the position of end quotes for multi-line docstrings"""
-    good_end = f"{(tokeninfo.start[1]) * ' '}{(tokeninfo.string[0]) * 3}"
-    new_string = tokeninfo.string
-    split_string = new_string.split("\n")
-
-    # Add new line with only quotes
-    if not new_string.endswith("\n" + good_end):
-        new_string = new_string[:-3] + "\n" + good_end
-    # Remove line with only quotes for potential single line string
-    elif len(split_string) == 2 and split_string[-1] == good_end:
-        new_string = "\n".join(split_string[:-1]) + tokeninfo.string[0] * 3
-
-    return tokenize.TokenInfo(
-        tokeninfo.type, new_string, tokeninfo.start, tokeninfo.end, tokeninfo.line
-    )
-
-
-def _format_closing_quotes(tokeninfo: tokenize.TokenInfo) -> tokenize.TokenInfo:
+
+    def _treat_string(self, new_string: str) -> str:
+        if new_string[3] == "\n":
+            new_string = re.sub(r"\n *", "", new_string, 1)
+        return new_string
+
+
+class ClosingQuotesFormatter(Formatter):
```

Comment:
Shouldn't this be a `StringFormatter` as well?

Reply:
``return self._format_multiline_ending_quotes(tokeninfo)`` returns a token

Reply:
It does, but only ever changes the `string`. It's a bad design that gets shown by this wonderful refactor! 😄

Reply:
Hmm but it needs the token and not the string to work. Maybe treat_string needs to take a token as input ?

Reply:
Or we could initialize the formatters for each loop and the token become a class attribute.

Reply:
Ah that makes sense I think.

Reply:
> Or we could initialize the formatters for each loop and the token become a class attribute.

That sounds like a large impact on performance. Would probably still be manageable, but I like how quick the tool runs at the moment.

Reply:
I've refactored treat_string, the result does look cleaner.